raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -963,49 +979,111 @@ private FetchResponseData tryCompleteFetchRequest(
+            int lastFetchedEpoch = request.lastFetchedEpoch();
+            LeaderState state = quorum.leaderStateOrThrow();
+            ValidatedFetchOffsetAndEpoch validatedOffsetAndEpoch = validateFetchOffsetAndEpoch(fetchOffset, lastFetchedEpoch);
 
-            if (state.updateReplicaState(replicaId, currentTimeMs, info.startOffsetMetadata)) {
-                onUpdateLeaderHighWatermark(state, currentTimeMs);
+            final Records records;
+            if (validatedOffsetAndEpoch.type() == ValidatedFetchOffsetAndEpoch.Type.VALID) {
+                LogFetchInfo info = log.read(fetchOffset, Isolation.UNCOMMITTED);
+
+                if (state.updateReplicaState(replicaId, currentTimeMs, info.startOffsetMetadata)) {
+                    onUpdateLeaderHighWatermark(state, currentTimeMs);
+                }
+
+                records = info.records;
+            } else {
+                records = MemoryRecords.EMPTY;
             }
 
-            return buildFetchResponse(Errors.NONE, info.records, Optional.empty(), state.highWatermark());
+            return buildFetchResponse(Errors.NONE, records, validatedOffsetAndEpoch, state.highWatermark());
+        } catch (Exception e) {
+            logger.error("Caught unexpected error in fetch completion of request {}", request, e);
+            return buildEmptyFetchResponse(Errors.UNKNOWN_SERVER_ERROR, Optional.empty());
         }
     }
 
     /**
      * Check whether a fetch offset and epoch is valid. Return the diverging epoch, which
      * is the largest epoch such that subsequent records are known to diverge.
      */
-    private Optional<OffsetAndEpoch> validateFetchOffsetAndEpoch(long fetchOffset, int lastFetchedEpoch) {
-        if (fetchOffset == 0 && lastFetchedEpoch == 0) {
-            return Optional.empty();
+    private ValidatedFetchOffsetAndEpoch validateFetchOffsetAndEpoch(long fetchOffset, int lastFetchedEpoch) {
+        if (log.startOffset() == 0 && fetchOffset == 0) {
+            if (lastFetchedEpoch != 0) {
+                logger.warn(
+                    "Replica sent a zero fetch offset ({}) but the last fetched epoch ({}) was not zero",
+                    fetchOffset,
+                    lastFetchedEpoch
+                );
+            }
+            return ValidatedFetchOffsetAndEpoch.valid(new OffsetAndEpoch(fetchOffset, lastFetchedEpoch));
         }
 
-        OffsetAndEpoch endOffsetAndEpoch = log.endOffsetForEpoch(lastFetchedEpoch)
-            .orElse(new OffsetAndEpoch(-1L, -1));
-        if (endOffsetAndEpoch.epoch != lastFetchedEpoch || endOffsetAndEpoch.offset < fetchOffset) {
-            return Optional.of(endOffsetAndEpoch);
+
+        Optional<OffsetAndEpoch> endOffsetAndEpochOpt = log
+            .endOffsetForEpoch(lastFetchedEpoch)
+            .flatMap(endOffsetAndEpoch -> {
+                if (endOffsetAndEpoch.epoch == lastFetchedEpoch && endOffsetAndEpoch.offset == log.startOffset()) {
```

Comment:
So basically we are trying to detect when the fetch is outside the range of the log, which means we will need to send a snapshot to the follower. Is that right?

One thought we don't have to do here is perhaps we should change the semantics of `endOffsetForEpoch` as exposed by `ReplicatedLog`. Maybe we should only return the end offset of an epoch when we know it with certainty.

Reply:
Note that the latest implementation changed a bit but this comment should still apply.

Yeah. I think we can't trust the `endOffsetAndEpoch` returned by `kafka.log.Log` when the epoch is smaller than any known epoch in the log. We know this is the case if `endOffsetForEpoch` returns `new OffsetAndEpoch(log.startOffset(), lastFetchedEpoch)`. If `endOffsetForEpoch` returns this value we overwrite it to the `oldestSnapshotId`. In other words if the leader has the following state:

oldestSnapshotId: endOffset = 10, epoch = 2
logStartOffset: 10
leaderEpochCache: (epoch = 4, startOffset = 10), ...

If the follower sends the Fetch request for offset 11 and last fetched offset 3, then the leader should return a diverging epoch of OffsetAndEpoch(offset=10, epoch=2). If we don't do this overwrite the leader would return OffsetAndEpoch(offset=10, epoch=3).